docs/design/cloud-init.rst

```diff
@@ -0,0 +1,260 @@
+cloud images that are configured to work with cloud init, we would like to gain
+the benefit of working with official images. Additionally, a user will be able
+to further configure a VM that is created out of a cloud-init enabled image by
+using cloud-init's user-data mechanism.
+
+Enabling cloud-init
+-------------------
+
+In order to enable snf-image's cloud-init mode, the image should have set the
+CLOUD_INIT image property.
+
+Proposed changes
+================
+
+Datasources
+-----------
+
+Since we want snf-image to be able to work stand-alone, without having to
+depend of external services, the software could make use of only 2 datasources:
+* NoCloud
+* None
+All the others expect to fetch data from an external entity (CDROM, Floppy,
+serial-port, Web-service).
+
+The "None" is the fallback datasource when no other can be selected. It is the
+equivalent of an empty datasource in that it provides an empty string as
+userdata and an empty dictionary as metadata. We could make use of this and let
+*snf-image-helper* inject all it's configuration to files under
+``/etc/cloud/cloud.cfg.d``. We could then allow the user to provide extra
+configuration data by using an new OS parameter (e.g. cloud_usedata) which's
+content would be injected into a file under ``/etc/cloud/cloud.cfg.d``. The
+only problem with this approach is that the user may only provide YAML user
+data, which is not fully compatible with Openstack. The nova API allows the
+user to provide user-data upon server creation. Those user data are available
+in the OpenStack meta-data service to be consumed by cloud-init. Since we are
+using *snf-image* to build a stack that exposes an OpenStack compute API, it's
+highly discouraged to have a, by design, incompatibility.
+
+This leaves us with the NoCloud option. This datasource allows us to provide
+user-data and meta-data to the instance through a number of different ways. The
+most suitable for *snf-image* is the ``/var/lib/cloud/seed/nocloud-net``
+directory. The datasource will expect to find the files ``meta-data`` and
+``user-data`` and optionally ``verdor-data`` as well as ``network-config``
+under this directory. *snf-image* could use a dedicated task to created those
+files and write this:
+```
+datasource_list [NoCloud]
+```
+to the cloud-init configuration. The various tasks could then perform their
+configuration by truncating the ``meta-data`` and ``vendor-data`` files or by
+appending files to the cloud-init configuration directory. By definition, the
+best place for *snf-image* to put the configuration would be the vendor-data:
+
+> Vendordata is data provided by the entity that launches an instance (for
+> example, the cloud provider). This data can be used to customize the image to
+> fit into the particular environment it is being run in.
+
+Unfortunately, older versions of cloud-init don't play well with
+``vendor-data``, which leaves us with ``meta-data`` and the files under
+``/etc/cloud/cloud.cfg.d``.
```

Comment:
Not really but I've tried it on the latest jessie (https://cdimage.debian.org/cdimage/openstack/archive/8.8.0-20170506/debian-8.8.0-openstack-amd64.raw) and strech (https://cdimage.debian.org/cdimage/openstack/9.0.2-20170623/debian-9.0.2-20170623-openstack-arm64.raw) image. If you add the following to the vendor-file:
```
$ cat /var/lib/cloud/seed/nocloud-net/vendor-data 
#cloud-config
hostname: test
```
and then boot the image. You check under the `/var/lib/cloud/instance` directory where the datasource's state resides:
```
# ls -l /var/lib/cloud/instance/
total 48
-rw-r--r-- 1 root root   51 Jun 23 13:15 boot-finished
-rw------- 1 root root  242 Jun 23 13:15 cloud-config.txt
-rw-r--r-- 1 root root   94 Jun 23 13:15 datasource
drwxr-xr-x 2 root root 4096 Jun 23 13:15 handlers
-r-------- 1 root root 7384 Jun 23 13:15 obj.pkl
drwxr-xr-x 2 root root 4096 Jun 23 13:15 scripts
drwxr-xr-x 2 root root 4096 Jun 23 13:15 sem
-rw------- 1 root root  196 Jun 23 13:15 user-data.txt
-rw------- 1 root root  538 Jun 23 13:15 user-data.txt.i
-rw------- 1 root root    4 Jun 23 13:15 vendor-data.txt
-rw------- 1 root root   29 Jun 23 13:15 vendor-data.txt.i
```
```
# ls -l /var/lib/cloud/instance/
total 52
-rw-r--r-- 1 root root   51 Jun 23 13:20 boot-finished
-rw------- 1 root root  242 Jun 23 13:20 cloud-config.txt
-rw-r--r-- 1 root root   88 Jun 23 13:20 datasource
drwxr-xr-x 2 root root 4096 Jun 23 13:20 handlers
-r-------- 1 root root 6260 Jun 23 13:20 obj.pkl
drwxr-xr-x 2 root root 4096 Jun 23 13:20 scripts
drwxr-xr-x 2 root root 4096 Jun 23 13:20 sem
-rw------- 1 root root  196 Jun 23 13:20 user-data.txt
-rw------- 1 root root  501 Jun 23 13:20 user-data.txt.i
-rw------- 1 root root   65 Jun 23 13:20 vendor-cloud-config.txt
-rw------- 1 root root   29 Jun 23 13:20 vendor-data.txt
-rw------- 1 root root  334 Jun 23 13:20 vendor-data.txt.i
```

Both have a `vendor-data.txt.i` and a `vendor-data.txt` file, but in the old image the content is this:
```
# cat vendor-data.txt.i 
#cloud-config
hostname: test
```
```
# cat vendor-data.txt
None
```
and in the stretch instance it looks like this:

```
# cat vendor-data.txt.i 
Content-Type: multipart/mixed; boundary="===============4969262898977750000=="
MIME-Version: 1.0
Number-Attachments: 1

--===============4969262898977750000==
MIME-Version: 1.0
Content-Type: text/cloud-config
Content-Disposition: attachment; filename="part-001"

#cloud-config
hostname: test

--===============4969262898977750000==--
```
```
# cat vendor-data.txt
#cloud-config
hostname: test
```

The one file should contain the raw vendor-data and the other the YAML configuration. For the user-data in both images is works as expected. I've tried it with CentOS and Fedora too and the results are the same. Newer images have newer cloud-init versions and work as expected.